src/compiler/transformers/classProperties.ts

```diff
@@ -0,0 +1,489 @@
+/*@internal*/
+namespace ts {
+    const enum ClassPropertySubstitutionFlags {
+        /**
+         * Enables substitutions for class expressions with static fields
+         * which have initializers that reference the class name.
+         */
+        ClassAliases = 1 << 0,
+    }
+    /**
+     * Transforms ECMAScript Class Syntax.
+     * TypeScript parameter property syntax is transformed in the TypeScript transformer.
+     * For now, this transforms public field declarations using TypeScript class semantics
+     * (where the declarations get elided and initializers are transformed as assignments in the constructor).
+     * Eventually, this transform will change to the ECMAScript semantics (with Object.defineProperty).
+     */
+    export function transformClassProperties(context: TransformationContext) {
+        const {
+            hoistVariableDeclaration,
+            endLexicalEnvironment,
+            resumeLexicalEnvironment
+        } = context;
+        const resolver = context.getEmitResolver();
+
+        const previousOnSubstituteNode = context.onSubstituteNode;
+        context.onSubstituteNode = onSubstituteNode;
+
+        let enabledSubstitutions: ClassPropertySubstitutionFlags;
+
+        let classAliases: Identifier[];
+
+        /**
+         * Tracks what computed name expressions originating from elided names must be inlined
+         * at the next execution site, in document order
+         */
+        let pendingExpressions: Expression[] | undefined;
+
+        /**
+         * Tracks what computed name expression statements and static property initializers must be
+         * emitted at the next execution site, in document order (for decorated classes).
+         */
+        let pendingStatements: Statement[] | undefined;
+
+        return chainBundle(transformSourceFile);
+
+        function transformSourceFile(node: SourceFile) {
+            if (node.isDeclarationFile) {
+                return node;
+            }
+            const visited = visitEachChild(node, visitor, context);
+            addEmitHelpers(visited, context.readEmitHelpers());
+            return visited;
+        }
+
+        function visitor(node: Node): VisitResult<Node> {

```

Comment:
This is unfortunately inefficient as it will result in a full walk of the AST. We use `TransformFlags` as a way to indicate whether there is any work to do on a subtree so as not to traverse the entire AST when there is no work to be done.

How this works is that we add a flag to `TransformFlags` to indicate that a subtree of the AST contains nodes that require transformation. We set these flags in *binder.ts* since that is the first full walk of the AST that we perform following parse. For example, we previously set the `AssertTypeScript` flag (in `computePropertyDeclaration` [above](https://github.com/microsoft/TypeScript/pull/30467/files#diff-7455dc48a52c4974a3e3671fcfb1dfb9R3384)) as a way to indicate that the TypeScript transformer had work to do on the node.

For this case, lets assume we add a `ContainsClassFields` flag to `TransformFlags` that we will set on a property declaration in a class. Once these flags have been set, it becomes a simple matter of adding the following line at the top of `visitor`:
```ts
if (!(node.transformFlags & TransformFlags.ContainsClassFields)) return node;
```

This allows us to skip visiting an entire subtree when there is no work to be done, which is necessary to avoid regressing the performance of the emitter.

Reply:
Added this transform flag and performed related early checks in new commit.